core-js/src/main/javascript/cdf-legacy/components/core.js

```diff
@@ -1009,6 +1009,41 @@ var UnmanagedComponent = BaseComponent.extend({
     query.fetchData(this.parameters, handler, errorHandler);
   },
 
+  /**
+   * @summary Resolves the correct queryDefinition.
+   * @description <p>Resolves the correct queryDefinition, always defaulting to undefined.</p>
+   *
+   * @param {Object|Function} queryDef The query definition object or a getter function.
+   * @return {Object} The resolved queryDefinition object.
+   * @private
+   */
+  _resolveQueryDefinition: function(queryDef) {
+    var queryDefinition;
+    if(_.isFunction(this.getQueryDefinition)) {
+      queryDefinition = this.getQueryDefinition();
+    }
+    return this._isValidQueryDefinition(queryDefinition) ?
+        queryDefinition : (_.isFunction(queryDef) ? queryDef() : queryDef);
+  },
+
+  _isValidQueryDefinition: function (queryDef) {
+    return _.isObject(queryDef) && !_.isArray(queryDef) && !_.isEmpty(queryDef);
+  },
+
+  /**
+   * @summary Gets the queryDefinition object or undefined.
+   * @description <p>Gets the queryDefinition object or undefined.</p>
+   *
+   * @return {Object} The queryDefinition Object.
+   */
+  getQueryDefinition: function() {
+    if(!!this.queryDefinition) {
+      return this.queryDefinition;
+    } else if (!!this.chartDefinition) {
+      return this.chartDefinition;
+    }
```

Comment:
Maybe also check for `this.trafficDefinition` because of the [TrafficComponent](https://github.com/webdetails/cdf/blob/master/pentaho-js/src/main/javascript/cdf/components/TrafficComponent.js#L63) implementation.
Can we simply return the first value found? e.g.:
`return this.queryDefinition || this.chartDefinition || this.trafficDefinition;`.
This feels like an abstract function that each component should override accordingly.